rust/kernel/sync/spinlock.rs

```diff
@@ -178,3 +178,126 @@ unsafe impl<T: ?Sized> Lock for SpinLock<T> {
         &self.data
     }
 }
+
+/// Safely initialises an [`IrqDisableSpinLock`] with the given name, generating a new lock class.
+#[macro_export]
+macro_rules! irqdisable_spinlock_init {
+    ($spinlock:expr, $name:literal) => {
+        $crate::init_with_lockdep!($spinlock, $name)
+    };
+}
+
+/// A spinlock that always disables interrupts when locked.
+///
+/// This is similar to [`SpinLock`], except that the [`SpinLock::lock`] method always disables
+/// interrupts, and the guard restores the state when dropped. That is,
+/// [`IrqDisableSpinLock::lock`] behaves exactly like [`SpinLock::lock_irqdisable`].
+///
+/// # Examples
+///
+/// ```
+/// # use kernel::prelude::*;
+/// # use kernel::sync::IrqDisableSpinLock;
+/// # use core::pin::Pin;
+///
+/// struct Example {
+///     a: u32,
+///     b: u32,
+/// }
+///
+/// // Function that acquires spinlock and disables interrupts while holding it.
+/// fn lock_irqdisable_example(value: &IrqDisableSpinLock<Example>) {
+///     let mut guard = value.lock();
+///     guard.a = 10;
+///     guard.b = 20;
+/// }
+///
+/// // Initialises a spinlock and calls the example functions.
+/// pub fn spinlock_example() {
+///     // SAFETY: `spinlock_init` is called below.
+///     let mut value = unsafe { IrqDisableSpinLock::new(Example { a: 1, b: 2 }) };
+///     // SAFETY: We don't move `value`.
+///     kernel::irqdisable_spinlock_init!(unsafe { Pin::new_unchecked(&mut value) }, "value");
```

Comment:
Should this be
```suggestion
///     kernel::irqdisable_spinlock_init!(unsafe { Pin::new_unchecked(&mut value) }, value);
```
?

Reply:
Not really. The second argument is a string the kernel uses to report issues detected via lockdep.